### PR TITLE
[data] allow custom batching in iter_batches

### LINKED ISSUE #54197: [data] Custom local shuffling batcher/sampler for Dataset.iter_*
### Description

Allow customized "batcher" to be used in `iter_torch_batches`, `iter_batches` etc. Note: this seems to be related to the `Sampler` classes in pytorch (https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler).

See also https://github.com/ray-project/ray/issues/51392

Currently when `local_shuffle_buffer_size` is set, a `ShuffleBatcher` is used to shuffle a local buffer to avoid expensive global shuffle. However, the implementation is rather specific. It will be great if users can provide their implementations.

Also, there is already a base class `BatcherInterface` defined, which can be exposed as "DeveloperAPI". The usage should be similar to how user customizes `FileMetadataProvider` in some read APIs.

### Use case

It will be more flexible when users have some custom batching/epoch logic and want to have full control of their shuffling in the pipeline. For example, some models may be more sensitive to shuffling and it's not easy to keep increasing `local_shuffle_buffer_size`. In that case user can implement their own batcher..

### PR DESCRIPTION
**Pull request by wingkitlee0**
## Why are these changes needed?

I would like to customize the batching of blocks in `iter_torch_batches()` APIs in data ingestion for model training. I considered the following two ways, but both have some shortcomings:
* wrap around `iter_batches()` with another iterator; OR
* expose the underlying non-public `BatcherInterface` class

For example, users want to split each block by group and wrote something like:
```python
def splitter(iterable: Iterable[Any]) -> Iterable[Any]:
    for batch in iterable:
        for _, grouped in batch.groupby("id"):
            yield grouped
```
Option 1 means `dataloader = splitter(ds.iter_batches(.., batch_size=None))`. It kind of works but technically we have two places that handle block-to-batch conversion (inside/outside iter_batches)

Option 2 requires users to write a OOP-style iterator (has_batch, next_batch), which may be overkill for simple iterator. Users have to verify both implementations work the same.

### Changes
A new `BatchingIteratorInterface` is defined, which consolidates the block to batch iterations. Existing OOP-style batcher (`BatcherInterface`) implementations stay the same and they are wrapped by the `BatcherBasedBlockIterator` class. Users provide a factory function to customize the batching logic:
```python
dataloader = ds.iter_torch_batches(..., batching_factory=create_splitter_fn)
```
where `create_splitter_fn` creates an `BatchingIteratorInterface` instance.

The underlying `blocks_to_batches()` (which does the actual block iteration) becomes
```python
yield from iterator.iter_batches(blocks)
```
where `iter_batches` consolidates the `has_batch`/`next_batch` calls into for-loops:
```python
def iter_batches(self, block_iter: Iterator[Block]) -> Iterator[Batch]:
    ...
    # Process each block and yield complete batches
    for block in block_iter:
        for batch in self.iter_from_block(block):
            yield ...
    else:
        # After all blocks, get final batches
        for batch in self.iter_leftovers():
            yield ...
```

The new `BatchingIteratorInterface` only requires overriding one `iter_from_block` method. Back to the original example, users can implement
```python
class MyBlockIterator(BatchingIteratorInterface):
    def __init__(self, col):
        self.col = col

    def iter_from_block(self, block: Block) -> Iterator[Block]:
       for _, df in block.groupby(self.col):
            yield df
```
This will ensure each batch contains only a single group.

Finally, users can use either OOP-style iterator or pythonic iterator in their factory method:
* OOP-style: subclass `BatcherInterface` and implement required methods. Wrap the batcher instance with `BatcherBasedBatchingIterator`
* pythonic iterator: subclass `BatchingIteratorInterface` directly and implement `iter_from_block` and optionally `iter_leftovers`

### TODO
* I would like BatchingIteratorInterface to be DeveloperAPI
* not sure what is the best way to expose this in the public APIs, e.g., `iter_torch_batches`. Currently I just call it `batching_factory=None`.
* tests (no ray task involved)
* 

## Related issue number

Closes  #54197

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run pre-commit jobs to lint the changes in this PR. ([pre-commit setup](https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#lint-and-formatting))
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wingkitlee0** [comment]
I will add tests...but it will be great to have some high-level review.

**Comment by xinyuangui2** [comment]
Hey thanks for the callout. We are thinking of moving batching and collate into ray data, and only keeping a thin layer inside iterator side. Will circle back soon.

**Comment by wingkitlee0** [comment]
Just curious if there is any update on this

Hopefully this PR isn't too big, which aims to allow the user to customize the batching logic. The aforementioned refactoring is related but does not seem to be a hard dependency (ideally this PR can go in earlier..)

**Comment by raulchen** [comment]
> Option 1 means dataloader = splitter(ds.iter_batches(.., batch_size=None)). It kind of works but technically we have two places that handle block-to-batch conversion (inside/outside iter_batches)

This looks fine to me. iter_batches is yielding batches. And spitter is consuming the batches with a group by, which should be transparent to iter_batches. 

Can you elaborate on your concern?

**Comment by wingkitlee0** [comment]
Stale. Closing it for now.
